### ISSUE TITLE
Can access Kong Admin API with nginx basic authen

### ISSUE THREAD

**Issue by vankhoa011** [PGBI/kong-dashboard#41]
Hello,

I setup Kong Admin API with nginx basic authentication. I saw kong-dashboard support login with basic authentication. However, It is not work.

Do you have any way to debug or investigate ?

**Comment by maestr0** [comment]
If I go to Settings and update Kong node URL and choose BasicAuth I'm getting an error 

Can't access Kong node with this url/port

I checked in the chrome web inspector if the UI sends the Authorization header and it does not.

**Comment by hnq90** [comment]
:+1: I got same error.

**Comment by PGBI** [comment]
@vankhoa011 can you please provide your nginx configuration file (the part w/ the basic auth)?

**Comment by maestr0** [comment]
I don't think the issue is on the nginx side.

I start the dashboard on my machine

> kong-dashboard start

open

http://localhost:8080/#/config

enter HTTPS url for the kong admin api `https://somepublicurl.example.com` and I choose `Basic Auth`.
The admin endpoint is configured correctly for BasicAuth. It works with CURL.

kong admin endpoint CORS headers
```
Access-Control-Allow-Credentials:true
Access-Control-Allow-Origin:*
Access-Control-Allow-Origin:*
Access-Control-Expose-Headers:Accept,Accept-Version,Content-Length,Content-MD5,Content-Type,Date,Content-Type,Authorization
```

I think the problem is with the dashboard UI http client. I checked the request headers of the first request that the UI makes when you hit SAVE button in the Config view. That request does not contain `Authorization` header.

I went even further. I debugged the code and at this point 

https://github.com/PGBI/kong-dashboard/blob/master/src/js/services/kong.js#L86



`$http.defaults.headers.common`  contains `Authorization` header...but it is not being used in the request. I don't know Angular enough to debug it further.

**Comment by PGBI** [comment]
I'm not saying the problem is in the nginx conf. I'd like to try to debug kong-dashboard but I am just too lazy to check how to configure basic auth on nginx :)

**Comment by maestr0** [comment]
What I did was I actually proxy the admin api via Kong itself with the basic auth plugin

Start Kong locally
Create a proxy for 

kong-admin.com -> localhost:8001

add 
```
kong-admin.com   127.0.0.1  
```
to   /etc/hosts

**Comment by vankhoa011** [comment]
@PGBI :  I configured as you mentioned on other issue. 

I added http://kong-api-gateway.com/kong and upstream to 127.0.0.1 with 8001, 
Enable CORS, then Add Basic Authentication Plugin. 

Then use Kong DashBoard with Basic Authentication. However, it is failed because The Kong Admin didn't allow OPTION action.

XMLHttpRequest cannot load  XXXX . Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin XXX is therefore not allowed access.

**Comment by maestr0** [comment]
I managed to solve the issue. Kong Dashboard code is correct. The problem is with the CORS plugin and the kong admin api.

This is what I did:
1 - create admin api
           kong-admin.mydomain.com -> localhost:8001
2 - create a consumer with BasicAuth creds
3 - enable CORS plugin on the admin api

Generally this should work but there is a bug/feature in the Kong Admin api. The api always sets `Access-Control-Allow-Origin:*` header regardless of the CORS plugin. You end up having two `Access-Control-Allow-Origin:*` headers and that causes problems. 
It is possible to disable the default Cors header on the admin api in kong.yml

Just comment out those three lines

```
location / {
        default_type application/json;
        content_by_lua_block {
#        ngx.header["Access-Control-Allow-Origin"] = "*"
          if ngx.req.get_method() == "OPTIONS" then
#           ngx.header["Access-Control-Allow-Methods"] = "GET,HEAD,PUT,PATCH,POST,DELETE"
#           ngx.header["Access-Control-Allow-Headers"] = "Content-Type"
             ngx.exit(204)
          end
          local lapis = require "lapis"
          lapis.serve "kong.api.app"
        }
```

**Comment by ahr73** [comment]
I've got the same problem (and I don't have a CORS plugin enabled)

When I didn't use basic-auth for admin API, dashboard worked perfectly.

1) I've modified kong.yml in the server admin-api part:

```
server {
      listen {{admin_api_listen}};
...
      location / {
        default_type application/json;

        # basic authentication for admin API
        auth_basic           "closed site";
        auth_basic_user_file /etc/kong/conf/htpasswd;

        content_by_lua_block {
          ngx.header["Access-Control-Allow-Origin"] = "*"
          if ngx.req.get_method() == "OPTIONS" then
            ngx.header["Access-Control-Allow-Methods"] = "GET,HEAD,PUT,PATCH,POST,DELETE"
            ngx.header["Access-Control-Allow-Headers"] = "Content-Type"
            ngx.exit(204)
          end
          local lapis = require "lapis"
          lapis.serve "kong.api.app"
        }
      }
...
```
htpasswd file contains just one user "admin".

When I try this configuration via SoapUI, it works correctly:
no authentication
response for
http://localhost:8001/apis
is  401 Authorization Required
It works fine, when I add basic-auth to the request

Kong Dashboard stop working with/without basic-auth setup. It's there just error message "Can't access a kong node with this url/port"

I've got both Kong even Kong Dashboard installed on localhost (VirtualBox with CentOS6)

**Comment by maestr0** [comment]
@ahr73 your issue is not related to Kong Dashboard. 

I don't know how you configured the admin api but it needs to return CORS headers otherwise Kong Dashboard (run from localhost:8080) is not going to work at all even without BasicAuth. 

```
Access-Control-Allow-Credentials:true
Access-Control-Allow-Origin:*
Access-Control-Expose-Headers:Accept,Accept-Version,Content-Length,Content-MD5,Content-Type,Date,Content-Type,Authorization
```

**Comment by kanikamnn** [comment]
Hi, I am facing the same issue, I have installed kong latest version 0.9.0 and kong dashboard.
When I am trying kong admin url without authentication its working perfectly fine.
But when I am configuring kong node url with basic auth submit button is not submitting any request.

P.S : I have secured my kong admin url by proxying it with kong and enabled basic auth plugin. This thing is working correctly when calling from any rest client.
